File: GoChat/src/main/java/Lib/msg/ChatMessage.java
package Lib.msg;

import java.util.HashMap;
import java.util.Map;

/**
 * 创建时间：
 * 更新时间 2017/11/1 23:05
 * 版本：
 * 作者：Mr.jin
 * 描述：聊天信息
 */

public class ChatMessage {
	private Map<String, Object> map;

	private MessageBody body;
	private String receiver;
	private String account;
	private String token;

	private Type type;

	public enum Type {
		TEXT, IMAGE, INVITATION
	}

	private ChatMessage() {
	}

	public static ChatMessage createMessage(Type type) {
		ChatMessage msg = new ChatMessage();
		msg.type = type;

		return msg;
	}

	public void setBody(MessageBody body) {
		this.body = body;
	}

	public MessageBody getBody() {
		return body;
	}

	public String getReceiver() {
		return receiver;
	}

	public void setReceiver(String receiver) {
		this.receiver = receiver;
	}

	public void setAccount(String account) {
		this.account = account;
	}

	public void setToken(String token) {
		this.token = token;
	}

	public Map<String, Object> getMap() {
		map = new HashMap<String, Object>();

		map.put("type", "request");
		map.put("sequence", SequenceCreater.createSequence());
		map.put("sender", account);
		map.put("token", token);
		map.put("action", getTypeString(type));
		map.put("receiver", receiver);
		map.put("content", getMessageBody(body));
		return map;
	}

	private String getTypeString(Type type) {
		switch (type) {
		case TEXT:
			return "text";
		case IMAGE:
			return "image";
		case INVITATION:
			return "invitation";
		default:
			break;
		}
		return null;
	}

	private String getMessageBody(MessageBody body) {
		if (body instanceof TextBody) {
			return ((TextBody) body).getContent();
		} else if (body instanceof InvitationBody) {
			return ((InvitationBody) body).getContent();
		}
		return null;
	}
}

File: chat/src/main/java/com/jw/chat/model/ChatMessage.kt
package com.jw.chat.model

import java.util.HashMap

/**
 * 创建时间：
 * 更新时间 2017/11/1 23:05
 * 版本：
 * 作者：Mr.jin
 * 描述：聊天信息
 */

class ChatMessage private constructor() {
    private var map: HashMap<String, Any>? = null

    var body: MessageBody? = null
    var receiver: String? = null
    private var account: String? = null
    private var token: String? = null

    private var type: Type? = null

    enum class Type {
        TEXT, IMAGE, INVITATION
    }

    fun setAccount(account: String) {
        this.account = account
    }

    fun setToken(token: String) {
        this.token = token
    }

    fun getMap(): Map<String, Any> {
        map = HashMap()
        map!!["type"] = "request"
        map!!["sequence"] = SequenceCreater.createSequence()
        map!!["sender"] = this.account!!
        map!!["token"] = this.token!!
        map!!["action"] = this.getTypeString(type!!)!!
        map!!["receiver"] = this.receiver!!
        map!!["content"] = this.getMessageBody(body)!!
        return map as HashMap<String, Any>
    }

    private fun getTypeString(type: Type): String? {
        when (type) {
            Type.TEXT -> return "text"
            Type.IMAGE -> return "image"
            Type.INVITATION -> return "invitation"
            else -> {
            }
        }
        return null
    }

    private fun getMessageBody(body: MessageBody?): String? {
        if (body is TextBody) {
            return body.content
        } else if (body is InvitationBody) {
            return body.content
        }
        return null
    }

    companion object {
        fun createMessage(type: Type): ChatMessage {
            val msg = ChatMessage()
            msg.type = type
            return msg
        }
    }
}